hoodie-client/src/main/java/com/uber/hoodie/func/LazyInsertIterable.java

```diff
@@ -56,13 +64,49 @@ public LazyInsertIterable(Iterator<HoodieRecord<T>> sortedRecordItr, HoodieWrite
   protected void start() {
   }
 
-
   @Override
   protected List<WriteStatus> computeNext() {
-    List<WriteStatus> statuses = new ArrayList<>();
+    final HoodieSparkTaskContext hoodieSparkTaskContext = HoodieSparkTaskContext.createNewHoodieSparkTaskContext();
+    List<WriteStatus> statuses = new LinkedList<>();
+    BufferedIterator<T, HoodieRecord<T>> bufferedIterator = new BufferedIterator<>(inputItr, hoodieConfig);
+    final Semaphore isDone = new Semaphore(0);
+    AtomicReference<Exception> writeError = new AtomicReference<>();
+    Thread writerThread =
+        new Thread(() -> {
+          try {
+            statuses.addAll(handleWrite(hoodieSparkTaskContext, bufferedIterator));
+          } catch (Exception e) {
+            logger.error("error writing hoodie records", e);
+            if (writeError.compareAndSet(null, e)) {
+              bufferedIterator.markAsFailed(e);
```

Comment:
tbh not a big fan of by hand threads and error notification thread-thread.. Can we just use ExecutorService an https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorCompletionService.html if needed, to fire off two futures and cancel one if the other fails.. 

the need for semaphore here should go away..